File: telegram-bot-service/src/main/java/telegram/polling/LongPollingUpdateAction.java
package telegram.polling;

import com.github.telegram.api.BotApi;
import com.github.telegram.api.Response;
import com.github.telegram.domain.Update;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.function.Consumer;

class LongPollingUpdateAction implements Runnable {

    private static final Logger log = LoggerFactory.getLogger(LongPollingUpdateAction.class);

    private static final int POLL_TIMEOUT_SEC = 60;
    private static final int POLL_ITEMS_BATCH_SIZE = 100;

    private final BotApi botApi;
    private final Consumer<Update> updateConsumer;
    private long lastUpdateOffset;

    LongPollingUpdateAction(BotApi botApi, Consumer<Update> updateConsumer) {
        this.botApi = botApi;
        this.updateConsumer = updateConsumer;
    }

    @Override
    public void run() {

        try {
            Response<List<Update>> response = botApi
                    .getUpdates(lastUpdateOffset, POLL_ITEMS_BATCH_SIZE, POLL_TIMEOUT_SEC);

                processUpdates(response);
        } catch (RuntimeException e) {
            log.error("Failed to call telegram api", e);
        }
    }

    public void processUpdates(Response<List<Update>> updateResponse) {
        if (!updateResponse.getOk() || updateResponse.getResult() == null) {
            log.error("Update failed, offset = {}", lastUpdateOffset);
            updateLastUpdateOffset(updateResponse.getResult());
            return;
        }

        List<Update> updates = updateResponse.getResult();

        updates.forEach(updateConsumer);

        // assuming that last update have highest offset (?)
        // todo: verify
        if (updates.size() > 0) {
            updateLastUpdateOffset(updates);
        }
    }

    private void updateLastUpdateOffset(List<Update> updates) {
        if (updates != null && updates.size() > 0) {
            Update lastUpdate = updates.get(updates.size() - 1);
            log.trace("offset: {} -> {}", lastUpdateOffset, lastUpdate.getUpdateId() + 1);
            lastUpdateOffset = lastUpdate.getUpdateId() + 1;
        }
    }

}

File: telegram-bot-bumblebee-core/src/main/java/com/github/bumblebee/polling/LongPollingUpdateAction.kt
package com.github.bumblebee.polling

import com.github.telegram.api.BotApi
import com.github.telegram.api.Response
import com.github.telegram.domain.Update
import org.slf4j.LoggerFactory

internal class LongPollingUpdateAction(private val botApi: BotApi,
                                       private val updateConsumer: (Update) -> Unit) : Runnable {
    private var lastUpdateOffset: Long = 0

    override fun run() {
        try {
            log.info("getUpdates with offset {}", lastUpdateOffset)
            val response = botApi.getUpdates(lastUpdateOffset, POLL_ITEMS_BATCH_SIZE, POLL_TIMEOUT_SEC)
            log.info("Got {} updates", response.result?.size)
            processUpdates(response)
        } catch (e: RuntimeException) {
            log.error("Failed to call telegram api", e)
        }
    }

    private fun processUpdates(response: Response<List<Update>>) {
        if (!response.ok || response.result == null) {
            log.error("Update failed, offset = {}", lastUpdateOffset)
            updateLastUpdateOffset(response.result)
            return
        }

        val updates = response.result!!

        updates.forEach(updateConsumer)

        if (updates.isNotEmpty()) {
            updateLastUpdateOffset(updates)
        }
    }

    private fun updateLastUpdateOffset(updates: List<Update>?) {
        if (updates != null && updates.isNotEmpty()) {
            val (updateId) = updates[updates.size - 1]
            log.trace("offset: {} -> {}", lastUpdateOffset, updateId + 1)
            lastUpdateOffset = updateId + 1
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(LongPollingUpdateAction::class.java)

        private val POLL_TIMEOUT_SEC = 60
        private val POLL_ITEMS_BATCH_SIZE = 100
    }
}
